### PR TITLE
Longhorn v1.2.0: Bump Kubernetes, API version group, CSI component's dependency version

### PR DESCRIPTION
**Pull request by PhanLe1010**
- [x] Bump Kubernetes, client-go to v1.18
- [x] API version group
- [x] CSI component's dependency version 

longhorn/longhorn#2757

### CODE REVIEW

File: csi/util.go
Comment (PhanLe1010) [inline csi/util.go:36]: Can you double-check this change @joshimoo ? Kubenetes v1.18 moves the mount package around and I have to make this change

**Comment by innobead** [comment]
This will be after RC1.

**Comment by PhanLe1010** [comment]
RKE `v1.18.20`
- [ ] Running e2e test
- [ ] Test upgrade from Longhorn `v1.1.2` to `v1.2.0`

K3s  `v1.22.1-rc1+k3s1`
- [ ] Running e2e test
- [ ] Test upgrade from Longhorn `v1.1.2` to `v1.2.0`

File: deploy/longhorn-images.txt
Comment (innobead) [inline deploy/longhorn-images.txt:5]: Why we don't use the latest releases of attacher (3.3.0), provisioner (v2.2.2), resizer (1.3.0) compatible with 1.18?

**Review by innobead — DISMISSED** [review]
In general, LGTM. Just one question to resolve.

Also, remember:
- update the example options of CSI images in longhorn-manager/deploy/install/02-components/04-driver.yaml
- have a PR update for longhorn/longhorn chart

Need @shuo-wu or @joshimoo to review before merging.

File: deploy/longhorn-images.txt
Comment (innobead) [reply deploy/longhorn-images.txt:5]: Using the upstream images is good, but this also possibly brings us another challenge when dealing with vulnerability issues the future. However, these images seem all based on `distroless` image, so that should be fine.

cc @meldafrawi

Comment (PhanLe1010) [reply deploy/longhorn-images.txt:5]: > Why we don't use the latest releases of attacher (3.3.0), provisioner (v2.2.2), resizer (1.3.0) compatible with 1.18?

So we import Kubernetes v1.18 (since many of our users are still using it). This leads to the decision of choosing CSI Spec v1.2 since Kubernetes v1.18 uses CSI Spec v1.2. As a result, all CSI components need to support min CSI Version v1.2

We cannot use provisioner (v2.2.2), resizer (1.3.0). But yeah, we can use attacher (3.3.0). I was looking at [the documentation](https://kubernetes-csi.github.io/docs/external-attacher.html) so I didn't see the latest attached v3.3.o

Comment (PhanLe1010) [reply deploy/longhorn-images.txt:5]: Just take another look at [attacher v3.3.0](https://github.com/kubernetes-csi/external-attacher/releases). Actually, it is too new and has recommended Kubernetes version of 1.22. Should we still want to use it?

Comment (innobead) [reply deploy/longhorn-images.txt:5]: Thanks, but it's minor for a patch update for now. All good.

Comment (innobead) [reply deploy/longhorn-images.txt:5]: Let's just keep the version u verified already.

**Comment by innobead** [comment]
> RKE `v1.18.20`
> 
> * [x]  e2e core test
> * [x]  Test upgrade from Longhorn `v1.1.2` to `v1.2.0`
> 
> K3s `v1.22.1-rc1+k3s1`
> 
> * [x]  e2e core test
> * [x]  ~Test upgrade from Longhorn `v1.1.2` to `v1.2.0`~ Not applicable since Longhorn `v1.1.2` doesn't run on Kubernetes v1.22

cc @khushboo-rancher @meldafrawi @kaxing

File: csi/util.go
Comment (joshimoo) [reply csi/util.go:36]: Looks good, god job on the change. With them moving this stuff into a dedicated util package we can potentially replace the nsenter exec after 1.2 and push some changes to the util package.

File: csi/util.go
Comment (joshimoo) [inline csi/util.go:33]: Can you also make the exec/mount changes on the share-manager repository, since it give you a chance to go through the build process there.

**Review by joshimoo — APPROVED** [review]
LGTM, @PhanLe1010 can you check if we still need the leader election patch on the new kubernetes version ref: https://github.com/longhorn/longhorn-manager/pull/639

**Comment by PhanLe1010** [comment]
> In general, LGTM. Just one question to resolve.
> 
> Also, remember:
> 
> * update the example options of CSI images in longhorn-manager/deploy/install/02-components/04-driver.yaml
> * have a PR update for longhorn/longhorn chart
> 
> Need @shuo-wu or @joshimoo to review before merging.

Done

File: csi/util.go
Comment (PhanLe1010) [reply csi/util.go:33]: Sure. Added to my todo list

**Comment by joshimoo** [comment]
The update for the share-manager was missed, so I am updating the share-manager kubernetes dependencies as part of
https://github.com/longhorn/longhorn/issues/2991

Which will go into v1.2.1

**Pull request merged**
